### ISSUE TITLE
Implement chunked uploading with lazy ops header

### ISSUE THREAD

**Issue by butonic** [owncloud/ocis#36]
We need to bring this to reva as clients want to drop support for the old chunking algorithm

**Comment by butonic** [comment]
In theory it would be possible to model the upload/quarantine area as a dedicated storage that as its last workflow step has the MOVE or if necessary COPY operation to the target storage. Unfortunately, this would need special handling of the fileid, because a cross storage move changes the file id. So when a client initiated an upload it will not get the correct file id until the file has reached the target storage...

how would this look like?
1. The client MKCOLs an upload dir. If a `Destination` header is present we can look up the responsible storage, if not we fall back to the users home storage. This translates to an InitiateUploadRequest, not a CreateCollectionRequest, because CS3 delegates upload to an out of bound process. Since we do not yet know the name we need to create a temporery file. If the client sent a destination, we may have a filename and can Initiate a direct upload to the target file.

2. The storage provider returns a location where we PUT the file. Currently, this is handled by the datasvc. This does however not need to use the same storage driver. It could use a new workflow driver that implements the execution of steps. The last step being: MOVE or COPY to the destination, when the client signals it has finished uploading. To decide on MOVE vs COPY the workflow storage can be configured with two storages: an ingress storage driver and a target storage driver. Both with a ... hmmm ... 

Intermezzo ...

The storage drivers only have an Upload function ... it currently does not allow append or range requests. This is a problem if we want to implement resumable uploads for CS3 and use HTTP/2 because we cannot resume uploads. Furthermore, the API does not allow implementing an assembly approach. As it stands, we need to have the file complete before we can call Upload.

If we assume the datasvc sits next to the storageprovidersvc they should have access to the storage. In the case of the owncloud and eos storage driver we could in theory bypass the Upload and directly write to the storage, including seeking to offsets etc. But for s3 this is not possible. S3 however does have multipart uploads. Does that map to our current chunked uploading? well, again, if we get the target in the initial MKCOL we can start the S3 multipart upload.

So, this is more a question of how to implement a datasvc that is storage driver specific. 
AFAICT https://github.com/tus/tusd/ can be embedded [programatically](https://github.com/tus/tusd/blob/master/docs/usage-package.md). We can implement different [DataStore implementations](https://godoc.org/github.com/tus/tusd/pkg#DataStore) to put the files where we need them. Hm, tusd already has a hook system: https://github.com/tus/tusd/blob/master/docs/hooks.md

... end intermezzo

- [ ] related: the reva gatewaysvc implementation for the storageprovider does not yet support MOVE https://github.com/cs3org/reva/blob/master/cmd/revad/svcs/grpcsvcs/gatewaysvc/storageprovidersvc.go#L179


I recommend we replace the datasvc with a tusd based implementation. But since  the current chunking does not send the taget in advance, and the CS3 api as well as the storage driver API only support uploading as a single operation we need to defer this and describe a proper solution before going further in that direction.

The current eos driver uses a temp file as well, so the owncloud driver, or any other for that matter has to suffer the penalty of doing a full copy of every file upload .... urgh this is :cow: :hankey:

**Comment by butonic** [comment]
Ok, so on the other half of the implementation we have an asynchronous chunk assembly. Normally I would use a proper task queue like https://github.com/RichardKnop/machinery/ to make the queu persistent, prevent multiple processes from starting the assembly ( locking or rether task deduplication).

However, let us remember the idea to push metadata as close to the storage as possible? S3 eg has its own multipart upload. I don't know if it allows uploads to the same upload id from different refgions ... I don't see a usecase for it, because a client is likely not going to switch between chunks. So do we need to trasport chunks over geo boundaries? If we want to allow asymettric download of chunks while the file is not fully uploaded that might be a good idea. But for this the chounks would need to be stored on the storage, not in the ocdav svc that currently implements chunking. So this needs to be postponed as well, until we discuss chunking v3?

Which leaves the question of how to provent multiple ocdavsvc processes from initiating an assembly at the same time. In a HA scenario we have at least two instances. Do they share a queue or do we lock the upload to prevent an additional assembly. Actually what if a client (or a proxy in between) repeats the move. We need to lock the chunked upload dir, not only to the process but to the go routing and deal with timeouts and errors ... tomorrow...

**Comment by butonic** [comment]
- [ ] wating for https://github.com/cs3org/reva/issues/284
- [ ] consider https://uppy.io in phoenix?

**Comment by refs** [comment]
blocked. see https://github.com/owncloud/ocis/issues/36#issuecomment-539488591

**Comment by butonic** [comment]
lazy ops got removed from the clients: https://github.com/owncloud/client/pull/8398
tracking new approach in https://github.com/owncloud/ocis/issues/214
